appcelerator/titanium_mobile
# feat: cross-platform light/dark mode API

**JIRA:** https://jira.appcelerator.org/browse/TIMOB-27501

**Description:**
This incorporates the code from @drauggres in his PR #11301

This attempts to make a more comprehensive cross-platform API for dealing with determining the current theme/mode, and consistency in the return value of `Ti.UI.fetchSemanticColor`.

**API**
- `Ti.UI.userInterfaceStyle` reports the current theme via `Ti.UI` constants: `USER_INTERFACE_STYLE_DARK`, `USER_INTERFACE_STYLE_LIGHT`, and `USER_INTERFACE_STYLE_UNSPECIFIED`
- `Ti.UI.fetchSemanticColor(String name)` returns a `Ti.UI.Color` proxy for the named color (and you can get the underlying "color" by calling `toHex()` on it to get a hex string) on iOS 13+, a string (representing the hex value) on Android and iOS < 13. The mismatched return types are necessary to make it easy to just pass the returned value to the UI properties without conversion right now.
  - Note that we support usage of the "standard" named colors baked into iOS that are adaptive to theme changes: [Standard Colors](https://developer.apple.com/documentation/uikit/uicolor/standard_colors?language=objc) and [UI Element Colors](https://developer.apple.com/documentation/uikit/uicolor/ui_element_colors?language=objc)
- `Ti.UI` now has a new `userInterfaceStyle` event which gets fired when the device theme changes. The object received in the event listener has a `value` property whose value is the constant for the new `USER_INTERFACE_STYLE_*` that was set.

**Deprecations**
- `Ti.App.iOS.userInterfaceStyle` -> use `Ti.UI.userInterfaceStyle`
- `Ti.App.iOS.USER_INTERFACE_STYLE_DARK` -> use `Ti.UI.USER_INTERFACE_STYLE_DARK`
- `Ti.App.iOS.USER_INTERFACE_STYLE_LIGHT` -> use `Ti.UI.USER_INTERFACE_STYLE_LIGHT`
- `Ti.App.iOS.USER_INTERFACE_STYLE_UNSPECIFIED` -> use `Ti.UI.USER_INTERFACE_STYLE_UNSPECIFIED`
- Undocumented `Ti.UI.iOS.fetchSemanticColor()` (use `Ti.UI.fetchSemanticColor()`)

**Doc updates**
- Document the `Ti.UI.Color` proxy. This is really an iOS-only proxy. We need a way to pass around the underlying `UIColor` objects from iOS so that we can set various UI color properties with them and they get "automatically" updated if the user changes the device theme.

**Notable parity issues**
While passing around `Ti.UI.Color` proxies on iOS allows for the UI to automatically adjust to theme changes, we don't have the equivalent on Android. Can we hook an equivalent `Ti.UI.Color` proxy that could auto-update on Android?

**To Test**
The actual API has some unit tests to verify the new properties/constants exist.

It's difficult to test on the emulator/sim, but to test the new event on theme changes you can run this app:
### Resources/semantic.colors.json ###
```json
{
	"primaryBackground": {
		"light": "#efece3",
		"dark": "#001900"
	}
}
```
### Resources/app.js ###
```javascript
// Also try baked-in colors on iOS like 'systemRedColor' - they will change as light/dark mode changes - although very subtly (I had to use the Digital Color Meter to double-check)
const primaryBackground = Ti.UI.fetchSemanticColor('primaryBackground');
const win = Ti.UI.createWindow({
    backgroundColor: primaryBackground
});

const btn = Ti.UI.createButton({
    title: 'Trigger'
});

btn.addEventListener('click', function() {
    Ti.API.info('Hello world!');
});

win.add(btn);
win.open();

Ti.UI.addEventListener('userinterfacestyle', e => {
    console.log(e);
    console.log(primaryBackground.toHex());
});
```
On iOS Emulator, I had to use the **Hardware > Siri** menu and then tell Siri to "Tun Dark mode on" (or off). On Android, You need an apilevel 29+ emulator, and you can drag from he top status bar down to get a quick menu. Click the pencil icon to edit the options displayed, drag the "Dark Theme" button up into it, and then you can use that to toggle dark/light themes while the app is running.
